homeassistant/components/minio/__init__.py

```diff
@@ -0,0 +1,280 @@
+        self.stop()
+
+
+class MinioListener:
+    """MinioEventThread wrapper with helper methods."""
+
+    def __init__(
+            self,
+            queue: Queue,
+            endpoint: str,
+            access_key: str,
+            secret_key: str,
+            secure: bool,
+            bucket_name: str,
+            prefix: str,
+            suffix: str,
+            events: List[str]
+    ):
+        """Create Listener."""
+        self.__queue = queue
+        self.__endpoint = endpoint
+        self.__access_key = access_key
+        self.__secret_key = secret_key
+        self.__secure = secure
+        self.__bucket_name = bucket_name
+        self.__prefix = prefix
+        self.__suffix = suffix
+        self.__events = events
+        self.__minio_event_thread = None
+
+    def start_handler(self, _):
+        """Create and start the event thread."""
+        from .minio_helper import MinioEventThread
+
+        self.__minio_event_thread = MinioEventThread(
+            self.__queue,
+            self.__endpoint,
+            self.__access_key,
+            self.__secret_key,
+            self.__secure,
+            self.__bucket_name,
+            self.__prefix,
+            self.__suffix,
+            self.__events
+        )
+        self.__minio_event_thread.start()
+
+    def stop_handler(self, _):
+        """Issue stop and wait for thread to join."""
+        if self.__minio_event_thread is not None:
+            self.__minio_event_thread.stop()
+
+
+def setup(hass, config):
+    """Set up MinioClient and event listeners."""
+    conf = config[DOMAIN]
+
+    host = conf[CONF_HOST]
+    port = conf[CONF_PORT]
+    access_key = conf[CONF_ACCESS_KEY]
+    secret_key = conf[CONF_SECRET_KEY]
+    secure = conf[CONF_SECURE]
+
+    queue_listener = QueueListener(hass)
+    queue = queue_listener.queue
+
+    hass.bus.listen_once(
+        EVENT_HOMEASSISTANT_START,
+        queue_listener.start_handler
+    )
+    hass.bus.listen_once(
+        EVENT_HOMEASSISTANT_STOP,
+        queue_listener.stop_handler
+    )
+
+    def _setup_listener(listener_conf):
+        bucket = listener_conf[CONF_LISTEN_BUCKET]
+        prefix = listener_conf[CONF_LISTEN_PREFIX]
+        suffix = listener_conf[CONF_LISTEN_SUFFIX]
+        events = listener_conf[CONF_LISTEN_EVENTS]
+
+        minio_listener = MinioListener(
+            queue,
+            get_minio_endpoint(host, port),
+            access_key,
+            secret_key,
+            secure,
+            bucket,
+            prefix,
+            suffix,
+            events
+        )
+
+        hass.bus.listen_once(
+            EVENT_HOMEASSISTANT_START,
+            minio_listener.start_handler
+        )
+        hass.bus.listen_once(
+            EVENT_HOMEASSISTANT_STOP,
+            minio_listener.stop_handler
+        )
+
+    for listen_conf in conf.get(CONF_LISTEN, []):
+        _setup_listener(listen_conf)
+
+    from minio import Minio
```

Comment:
We can now import requirements at the top of the module as a result of manifest.json.